### PR TITLE
Implement more lexer diagnostics

### PR DESCRIPTION
**Pull request by ahoppen**
This implements a few more lexer diagnostics.

Each commit of this PR should be standalone and should be reviewable on its own. But since they all build on top of each other, I can’t really create separate PRs for them.

### CODE REVIEW

**Comment by ahoppen** [comment]
@swift-ci Please test

File: Sources/SwiftSyntax/LexerError.swift [hunk 1]
```diff
@@ -46,6 +45,7 @@ public struct LexerError: Hashable {
   public init(_ kind: Kind, byteOffset: Int) {
     assert(byteOffset >= 0)
     // `type(of: self.byteOffset).max` gets optimized to a constant
+    assert(byteOffset >= 0)
```
Comment (bnbarham) [inline Sources/SwiftSyntax/LexerError.swift:61]: Duplicated?

**Review by bnbarham — COMMENTED** [review]
Nice!

File: Sources/SwiftParserDiagnostics/ParserDiagnosticMessages.swift [hunk 2]
```diff
@@ -474,6 +474,9 @@ extension FixItMessage where Self == StaticParserFixIt {
   public static var replaceCurlyQuoteByNormalQuote: Self {
     .init(#"replace curly quotes by '"'"#)
   }
+  public static var replaceNonBreakingSpaceBySpace: Self {
+    .init("replace non-breaking space by ' '")
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply Sources/SwiftParserDiagnostics/ParserDiagnosticMessages.swift:478]: Interesting. I would have always said `by` but Google agrees that `with` is preferred. This also changes the `ReplaceTokensFixIt`, so I’ll do it in a follow-up PR.

File: Sources/SwiftParser/Lexer/Cursor.swift [hunk 3]
```diff
@@ -287,8 +287,11 @@ extension Lexer.Cursor {
     // Leading trivia.
     let leadingTriviaStart = self
     let newlineInLeadingTrivia: NewlinePresence
+    var error: LexerError? = nil
     if let leadingTriviaMode = self.currentState.leadingTriviaLexingMode(cursor: self) {
-      newlineInLeadingTrivia = self.lexTrivia(mode: leadingTriviaMode)
+      let triviaResult = self.lexTrivia(mode: leadingTriviaMode)
+      newlineInLeadingTrivia = triviaResult.newlinePresence
+      error = error ?? triviaResult.error.map { LexerError($0.kind, byteOffset: cursor.distance(to: $0.position)) }
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply Sources/SwiftParser/Lexer/Cursor.swift:294]: Good idea. Did that.

File: Sources/SwiftSyntax/LexerError.swift (hunk 1 above)
Comment (ahoppen) [reply Sources/SwiftSyntax/LexerError.swift:66]: Thanks. Probably crept in due to a merge conflict.

File: Tests/SwiftParserTest/LexerTests.swift [hunk 4]
```diff
@@ -1001,4 +1005,100 @@ public class LexerTests: XCTestCase {
...
+  }
+
+  func testInvalidUnicodeEscapeSequence() {
+    AssertLexemes(
+      #"""
+      "1️⃣\u"
+      """#,
+      lexemes: [
+        LexemeSpec(.stringQuote, text: #"""#),
+        LexemeSpec(.stringSegment, text: #"\u"#, error: "expected hexadecimal code in braces after unicode escape"),
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply Tests/SwiftParserTest/LexerTests.swift:1048]: Good idea. I just copied error messages from the old lexer but your suggestion is definitely better.

File: Sources/SwiftParser/Lexer/Cursor.swift [hunk 5]
```diff
@@ -335,9 +351,7 @@ extension Lexer.Cursor {
       flags.insert(.isAtStartOfLine)
     }
 
-    let error = result.error.map { error in
-      return LexerError(error.kind, byteOffset: cursor.distance(to: error.position))
-    }
+    error = error ?? result.error.map { LexerError($0.kind, byteOffset: cursor.distance(to: $0.position)) }
```
Comment (bnbarham) [inline Sources/SwiftParser/Lexer/Cursor.swift:354]: Could we add a conversion method on `LexingError` that takes the `cursor`?
```
struct LexingError {
  ...
  func toLexerError(_ initialCursor: Cursor) {
    return LexerError(kind, initialCursor.distance(to: position)
  }
}
...
error = error ?? result.error?.toLexerError(cursor)
```

**Comment by ahoppen** [comment]
https://github.com/apple/swift/pull/63372

@swift-ci Please test

**Comment by ahoppen** [comment]
https://github.com/apple/swift/pull/63372

@swift-ci Please test

**Pull request merged**